### ISSUE TITLE
Generate fitness, health and wellness data.

### ISSUE THREAD

**Issue by anthonyroach** [synthetichealth/synthea#291]
I'm interested in generating synthetic data representing a person's self reported and device collected fitness, health and wellness data such as calories burned, exercise duration, heart rate, and intermittent fasting duration.  My interest is driven by the need for health and wellness service providers (like my employer) to have demo and test data for their services and products. 

I think a good first stab at this would be implementing a "Fitness" or "Wellness" module that generates Observations as described here:

http://www.healthintersections.com.au/?p=2487

The purpose of this issue is really to gauge interest in this sort of thing and see whether a module like this would be accepted by the maintainers, and also to gather feedback on the idea. I plan to implement this myself.

**Comment by dehall** [comment]
This sounds like a great idea and we would definitely welcome a pull request.

A few thoughts:
- Currently Synthea requires all Observations to be part of an Encounter, even though in this case there wouldn't be a real encounter. A quick workaround would be to add a "dummy" encounter, but then Synthea would associate that to a provider, which probably isn't what we want. I don't see this as a blocker but we can discuss further as necessary. 
- How much data would you expect to generate? Fitbits can produce tons of heart rate data points and make nice graphs from that -- we haven't done anything close to that volume and level of granularity but it would be interesting to hear your findings.
- Producing realistic but varied vital signs, such as heart rate, is likely to be a challenge in the current iteration of the Generic Module Framework. We'd like to add some capability to use expressions (#189) in the GMF but haven't settled on a final approach yet.

**Comment by anthonyroach** [comment]
Thanks for the feedback, and glad to hear that you like the idea.

- I hadn't realized that an Encounter is required. Certainly creating a "dummy" encounter would work in the short term. Is removing the requirement on having an Encounter something you guys are open to as a separate issue?
- I wasn't planning to generate massive amounts of data like a Fitbit would produce, at least not initially. I was thinking more like sampled values or average values (e.g. resting heart rate each day, and heart rate during exercise). For FHIR it looks like the correct way to model high volume or continuous data from a device is to have a single Observation for some reasonable time period and then have the value be a [SampleData](https://www.hl7.org/fhir/datatypes.html#SampledData) which contains an array of numbers (encoded as a string, interestingly enough). I would imagine supporting SampleData would be more involved than what I'd like to bite off right now.
- I'm still getting my brain around the GMF, but I think I can get pretty far with the existing vital-signs and attributes along with delays and states and loops. The state machine will be complex, and having math expressions of some sort would make things a lot easier, but I think I can get something basic going with the existing facilities. 

To summarize, my hope is I can get something basic working within the existing GMF, and then perhaps longer term look at helping to extend the GMF and adding more complexity and realism.

**Comment by jawalonoski** [comment]
Here is an idea that I'm not committing to:

- Add a `Device` state along with an `assign_to_attribute` property (this would become a FHIR Device on FHIR export).
- Add an optional property to the `Observation` state called `device` that references a device assigned to an attribute or identified by code.
- Add a way to include sampled data on `Observation`.

**Comment by jawalonoski** [comment]
Regarding math expressions, please see https://github.com/synthetichealth/synthea/issues/189

**Comment by tiloc** [comment]
I am working in the field of chronic care/telemedicine and would also be motivated to contribute module improvements for congestive heart failure and COPD. So far my attempts to model this as shown in the attached image have not yielded success. I was hoping that this setup would lead to a daily blood pressure reading added to the patient record until they suffer a flare, but I am currently not seeing any values added to the record? I think the idea outlined by @jawalonoski would match the use-case very well.

![image](https://user-images.githubusercontent.com/16414047/48671403-b18f2380-eb27-11e8-8756-5a9cafd012ca.png)

**Comment by tiloc** [comment]
I have found an encounter type that works and embedded the self-measurements into it. There is oftentimes a telemedical center with a nurse involved, so this is somewhat realistic. I am not seeing the expected frequency of measurements, though. I am getting a measurement approx. 1/month, not the expected 1/day. Any way how I can increase the frequency?
![image](https://user-images.githubusercontent.com/16414047/48671916-5e20d380-eb2f-11e8-82a8-e13df862496f.png)

**Comment by jawalonoski** [comment]
Well, I'm not 100% certain but my immediate theory is that this is related to the length of the default timestep combined with the Delay state. The default timestep is 7 days. The engine is supposed to _"go back in time"_ between the timesteps and fill in the missing moments, if necessary, but perhaps this isn't happening as it should... If you want to try your module with a daily timestep, you need to edit `src/main/resources/synthea.properties`:

```properties
generate.timestep = 604800000
# time is in ms
# 1000 * 60 * 60 * 24 * 7 = 604800000
```

For a single day, you'll want to set the value to `1000 * 60 * 60 * 24 = 86400000`.

Let me know if this helps or not. If it does, we probably have a bug in the _"go back in time"_ code.

**Comment by jawalonoski** [comment]
Yes, vital signs only get calculated once per timestep. You can increase the rate of calculation with the `generate.timestep` property. 

The code that sets the vital-sign values is here:
https://github.com/synthetichealth/synthea/blob/master/src/main/java/org/mitre/synthea/modules/LifecycleModule.java#L558

It currently is not set in the generic modules framework.

**Comment by jawalonoski** [comment]
The trend idea is interesting. I think it would be a big overhaul of how we generate vital signs today. I'm willing to consider a design proposal if you're interested in making such a contribution. I'd prefer you submit a proposal on something like that first, that we can iterate on, before you (or anyone else) spends a lot of time programming a change that we might have trouble merging.

**Comment by tiloc** [comment]
I am experimenting with an idea to have a better feel for its impact. I have made a few modifications to no longer map vital signs to doubles, but to a generator function that takes a time stamp as input. That should give me the ability to obtain a value at any specified time, be it a constant value, a random value, a trending value, something sophisticated for that vital sign, etc.
I think the idea might work, but my implementation is breaking unit tests, I am trying to understand why. Some it seems, because people are born without a blood pressure? I think there are some unit tests where a person is born and then their blood pressure is immediately recorded as an observation, but the LifeCycleModule never got to do its first .process, so the blood pressure never got set? Should an invocation of diabeticVitalsigns be added to the birth() method to give the newborn a blood pressure and some other basic physiological parameters?

**Comment by jawalonoski** [comment]
> Some it seems, because people are born without a blood pressure? I think there are some unit tests where a person is born and then their blood pressure is immediately recorded as an observation, but the LifeCycleModule never got to do its first .process, so the blood pressure never got set?

Yes, it looks like it.

>Should an invocation of diabeticVitalsigns be added to the birth() method to give the newborn a blood pressure and some other basic physiological parameters?

Yes, this seems reasonable.

**Comment by tiloc** [comment]
@jawalonoski see PR #448 for an illustration / design proposal of what I currently have in mind. It is a mechanism which can stitch together a long-term graph from a sequence of short-term trends. I think I will continue to test-drive this for the heart failure use-case. I have implemented it for BP, but weight would be another obvious candidate. I am also thinking about ways how a GMF could influence this trend generator. My use-case would be exacerbations or fast-acting treatments. In the case of HF, patients can suddenly gain >5 lb of weight overnight, go to the hospital and are brought back to their target dry-weight through treatment.

**Comment by jawalonoski** [comment]
@tiloc Thank you for the WIP -- we appreciate contributions. I looked at the PR (but I have not yet reviewed the code itself) -- and the description and illustration seems reasonable.
